File: send_iris.c
#include "iris.h"

// make iris.o happy
void iris_call_submit_result(struct pdu *pdu) { }
int iris_call_recv_data(int fd) { return 0; }
int iris_call_register_fd(int fd) { return 0; }
struct {
	char *host;
	int   port;
	int   timeout;
	int   quiet;
	char  delim[2];
} OPTS = {
	.host    = NULL,
	.port    = 0,
	.timeout = 10,
	.quiet   = 0,
	.delim   = "\t"
};

void alarm_handler(int sig)
{
	printf("Timed out after %d seconds\n", OPTS.timeout);
	alarm(0); exit(3);
}

int process_args(int argc, char **argv)
{
	int c;

	while ((c = getopt(argc, argv, "qvH:p:t:h?")) != -1) {
		switch (c) {
		case 'q':
			OPTS.quiet = 1;
			break;

		case 'v':
			OPTS.quiet = 0;
			break;

		case 'H':
			free(OPTS.host);
			OPTS.host = strdup(optarg);
			break;

		case 'p':
			OPTS.port = atoi(optarg);
			break;

		case 't':
			OPTS.timeout = atoi(optarg);
			break;

		case 'h':
		case '?':
			printf("USAGE: send_iris -H <host> [-p <port>] [-t <timeout>]\n");
			printf("\n");
			printf("  -h\n");
			printf("      Show this informative help screen.\n");
			printf("\n");
			printf("  -H <hostname>\n");
			printf("      IP or hostname of who we should submit results to.\n");
			printf("      (this option is required)\n");
			printf("\n");
			printf("  -p <port>\n");
			printf("      TCP port to connect to.\n");
			printf("      Defaults to 5668\n");
			printf("\n");
			printf("  -t <timeout>\n");
			printf("      Connection timeout, in seconds.\n");
			printf("      Defaults to %d\n", 10);
			printf("\n");
			exit(0);
			break;
		}
	}

	if (!OPTS.host) {
		fprintf(stderr, "Missing required -H option\n");
		return 1;
	}

	if (OPTS.port == 0)
		OPTS.port = 5668;
	return 0;
}

int main(int argc, char **argv)
{
	struct pdu *packets = NULL;
	int npackets = 0, nsent = 0;
	int sock, i;
	time_t now;

	if (process_args(argc, argv) != 0)
		exit(3);

	npackets = read_packets(stdin, &packets, OPTS.delim);

	signal(SIGALRM, alarm_handler);
	alarm(OPTS.timeout);
	if ((sock = net_connect(OPTS.host, OPTS.port)) < 0) {
		fprintf(stderr, "error connecting to %s:%d: %s\n",
				OPTS.host, OPTS.port, strerror(errno));
		alarm(0); exit(3);
	}

	for (i = 0; i < npackets; i++) {
		time(&now); packets[i].ts = (uint32_t)now;
		pdu_pack(&packets[i]);

		if (pdu_write(sock, (uint8_t*)&packets[i]) < 0) {
			fprintf(stderr, "error sending data to %s:%d\n", OPTS.host, OPTS.port);
			close(sock);
			alarm(0); exit(3);
		}
		nsent++;
	}
	shutdown(sock, SHUT_WR);
	fd_sink(sock);
	close(sock);
	alarm(0);

	if (!OPTS.quiet) {
		printf("Sent %d results to %s:%d\n", nsent, OPTS.host, OPTS.port);
	}

	free(packets);
	return 0;
}

File: send_iris.pl
#!/usr/bin/perl

#
# send_iris - IRIS Client Driver
#
# author:  James Hunt <jhunt@synacor.com>
# created: 2013-04-22
#

use strict;
use warnings;

use IO::Socket;
use Getopt::Long;
my %OPTIONS = (
	timeout   => 10,
	port      => 5668,
	delimiter => "\t",
);
GetOptions(\%OPTIONS,
	'help|h|?',
	'quiet|q',

	'host|H=s',
	'timeout|t=i',
	'port|p=i',
	'delimiter|d=s',

	'debug|D',

	'config|c=s',
);
if (!$OPTIONS{host}) {
	print STDERR "Missing required -H option\n";
	unless (-t STDIN) { while (<STDIN>) { } }
	exit 1;
}

my @CRC32 = ();

{
	for (my $i=0; $i<256; $i++) {
		my $x = $i;
		for (my $j=0; $j<8; $j++) {
			$x = ($x & 1) ? ($x >> 1) ^ 0xedb88320 : $x >> 1;
		}
		$CRC32[$i] = $x;
	}
};

sub crc32
{
	my ($s) = @_;
	my $crc = 0xffffffff;
	for (unpack 'C*', $s) {
		$crc = (($crc >> 8) & 0x00ffffff) ^ $CRC32[($crc ^ $_) & 0xff];
	}
	$crc ^ 0xffffffff;
}

$/ = "\x17";
my @packets = ();
my $pdu;
while (<STDIN>) {
	s/^[\x17\s]+//; s/[\x17\s]+$//;
	next if m/^$/;

	my ($host, $service, $rc, $output) = split /\s*$OPTIONS{delimiter}\s*/, $_, 4;
	if (!$output and $service and $service =~ m/^[0-9]$/) {
		$output  = $rc;
		$rc      = $service;
		$service = 'HOST';
	}
	$rc = 3 unless $rc >= 0 and $rc <= 3;
	next unless $host and $service;

	my $ts = time;
	$pdu = pack("NNnnZ64Z128Z4096", 0,    $ts, 1, $rc, $host, $service, $output);
	my $crc = crc32($pdu);
	substr($pdu, 0, 4) = pack("N", $crc);
	push @packets, $pdu;

	if ($OPTIONS{debug}) {
		print sprintf("PDU DUMP (crc32=%04x):\n", $crc);
		open my $od, "|-", "/usr/bin/od -a"
			or die "Failed to exec `od -a' for PDU debugging: $!\n";
		print $od $pdu;
		close $od;
	}
}

my $peer = "$OPTIONS{host}:$OPTIONS{port}";
my $n = 0;
my $socket = IO::Socket::INET->new(
	PeerAddr => $peer,
	Timeout  => $OPTIONS{timeout},
	Proto    => "tcp",
	Blocking => 1, # clients should *ALWAYS* block
) or die "Connect to $peer failed: $@\n";

for (@packets) {
	syswrite $socket, $_, length($_);
	$n++;
}
shutdown $socket, 1;

print STDERR "Sent $n results to $peer\n" unless $OPTIONS{quiet};
